gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/FeatureConverter.kt

```diff
@@ -0,0 +1,144 @@
+ */
+package com.google.android.gnd.persistence.remote.firestore.schema
+
+import kotlin.Throws
+import com.google.android.gnd.persistence.remote.DataStoreException
+import com.google.android.gnd.model.Project
+import com.google.android.gnd.model.feature.*
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.android.gnd.persistence.remote.firestore.schema.FeatureDocument
+import com.google.android.gnd.persistence.remote.firestore.schema.FeatureConverter
+import com.google.firebase.firestore.GeoPoint
+import timber.log.Timber
+import com.google.android.gnd.persistence.remote.firestore.schema.AuditInfoNestedObject
+import com.google.android.gnd.persistence.remote.firestore.schema.AuditInfoConverter
+import com.google.common.collect.ImmutableList
+import java8.util.Objects
+
+/** Converts between Firestore documents and [Feature] instances.  */
+object FeatureConverter {
+    const val LAYER_ID = "layerId"
+    const val LOCATION = "location"
+    const val CREATED = "created"
+    const val LAST_MODIFIED = "lastModified"
+    const val GEOMETRY_TYPE = "type"
+    const val POLYGON_TYPE = "Polygon"
+    const val GEOMETRY_COORDINATES = "coordinates"
+    const val GEOMETRY = "geometry"
+
+    @JvmStatic
+    @Throws(DataStoreException::class)
+    fun toFeature(project: Project, doc: DocumentSnapshot): Feature<*> {
+        val f = DataStoreException.checkNotNull(
+            doc.toObject(
+                FeatureDocument::class.java
+            ),
+            "feature data"
+        )
+
+        if (f.geometry != null && hasNonEmptyVertices(f)) {
+            return toFeatureFromGeometry(project, doc, f)
+        }
+
+        f.geoJson?.let {
+            val builder = GeoJsonFeature.newBuilder().setGeoJsonString(it)
+            fillFeature(builder, project, doc.id, f)
+            return builder.build()
+        }
+
+        f.location?.let {
+            val builder = PointFeature.newBuilder().setPoint(toPoint(it))
+            fillFeature(builder, project, doc.id, f)
+            return builder.build()
+        }
+
+        throw DataStoreException("No geometry in remote feature " + doc.id)
+    }
+
+    private fun hasNonEmptyVertices(featureDocument: FeatureDocument): Boolean {
+        val geometry = featureDocument.geometry
+
+        if (geometry == null
+            || geometry[GEOMETRY_COORDINATES] == null
+            || geometry[GEOMETRY_COORDINATES] !is List<*>
+        ) {
+            return false
+        }
+
+        val coordinates = geometry[GEOMETRY_COORDINATES] as List<*>?
+        return coordinates?.isNotEmpty() ?: false
+    }
+
+    private fun toFeatureFromGeometry(
+        project: Project, doc: DocumentSnapshot, f: FeatureDocument
+    ): PolygonFeature {
+        val geometry = f.geometry
+        val type = geometry!![GEOMETRY_TYPE]
+        if (POLYGON_TYPE != type) {
+            throw DataStoreException("Unknown geometry type in feature " + doc.id + ": " + type)
```

Comment:
Please use $var for concatenation. Same comment for the rest of the code as well.